File: tests/test-percpu.c
#include <stdlib.h>
#include <unistd.h>

#define NR_CPUS 32
#define __section(S) __attribute__ ((__section__(#S)))
#define DEFINE_PER_CPU(type, name) \
	__section(.data..percpu) __typeof__(type) name

/* Userspace blows up if you have symbols at offset 0, so we fake it */
#define per_cpu_offset(x) (__per_cpu_offset[x] - (unsigned long)&__per_cpu_start)

# define RELOC_HIDE(ptr, off)                                   \
  ({ unsigned long __ptr;                                       \
     __ptr = (unsigned long) (ptr);                             \
    (typeof(ptr)) (__ptr + (off)); })

#define SHIFT_PERCPU_PTR(__p, __offset)                                 \
        RELOC_HIDE((typeof(*(__p)) *)(__p), (__offset))

#define per_cpu_ptr(ptr, cpu)                                           \
({                                                                      \
        SHIFT_PERCPU_PTR((ptr), per_cpu_offset((cpu)));                 \
})

#define raw_per_cpu_ptr(ptr, cpu)        SHIFT_PERCPU_PTR((ptr), per_cpu_offset((cpu)))
#define raw_cpu_write(ptr, cpu, val) 		\
({						\
	*raw_per_cpu_ptr((ptr), cpu) = val;	\
})

struct test_struct {
	int x;
	unsigned long ulong;
	void *ptr;
};

unsigned long __per_cpu_offset[NR_CPUS];

DEFINE_PER_CPU(struct test_struct, struct_test);
DEFINE_PER_CPU(unsigned long, ulong_test);
DEFINE_PER_CPU(void *, voidp_test);
DEFINE_PER_CPU(struct test_struct *, ptr_to_struct_test);
DEFINE_PER_CPU(unsigned long *, ptr_to_ulong_test);

extern unsigned long __per_cpu_start;
extern unsigned long __per_cpu_end;
extern unsigned long __per_cpu_load;

struct test_struct *percpu_test;
struct test_struct *non_percpu_test;

int
main(void)
{
	int i;
	unsigned long size = (void *)&__per_cpu_end - (void *)&__per_cpu_start;

	for (i = 0; i < NR_CPUS; i++)
	{
		int ret;
		struct test_struct *f;
		void *ptr;
		unsigned long *l;

		ret = posix_memalign(&ptr, 4096, size);
		if (ret)
			return 1;

		__per_cpu_offset[i] = (unsigned long)ptr;

		f = per_cpu_ptr(&struct_test, i);
		f->x = i;
		f->ulong = i;
		f->ptr = NULL;

		raw_cpu_write(&ulong_test, i, i);
		raw_cpu_write(&voidp_test, i, (void *)0xdeadbeef);
		raw_cpu_write(&ptr_to_struct_test, i, f);
		raw_cpu_write(&ptr_to_ulong_test, i, &f->ulong);
	}

	percpu_test = &struct_test;
	non_percpu_test = per_cpu_ptr(&struct_test, 0);

	/* We want to give gdb a core dump to work with */
	abort();
	return 0;
}

File: tests/test_percpu.py
# -*- coding: utf-8 -*-
# vim:set shiftwidth=4 softtabstop=4 expandtab textwidth=79:

import unittest
import gdb

import crash
import crash.types.percpu as percpu

class TestPerCPU(unittest.TestCase):
    def setUp(self):
        gdb.execute("file tests/test-percpu", to_string=True)

        try:
            print()
            print("--- Unsuppressable gdb output ---", end='')

            gdb.execute("run", to_string=False)
            msym = gdb.lookup_minimal_symbol('__per_cpu_load')
            self.baseaddr = msym.value().address
            self.test_struct = gdb.lookup_type("struct test_struct")
            self.ulong_type = gdb.lookup_type("long unsigned int")
            self.voidp = gdb.lookup_type("void").pointer()
        except gdb.error as e:
            # If we don't tear it down, the rest of the tests in
            # other files will fail due to it using the wrong core file
            self.tearDown()
            raise(e)

    def tearDown(self):
        try:
            gdb.execute("detach", to_string=True)
            gdb.execute("file")
        except gdb.error:
            print()
            pass
        print("--- End gdb output ---")

    def test_struct_test(self):
        var = gdb.lookup_symbol('struct_test', None)[0]
        self.assertTrue(var is not None)
        for cpu, val in list(percpu.get_percpu_vars(var).items()):
            self.assertTrue(val['x'] == cpu)
            self.assertTrue(val.type == self.test_struct)

    def test_ulong_test(self):
        var = gdb.lookup_symbol('ulong_test', None)[0]
        self.assertTrue(var is not None)
        for cpu, val in list(percpu.get_percpu_vars(var).items()):
            self.assertTrue(val == cpu)
            self.assertTrue(val.type == self.ulong_type)

    def test_ulong_ptr_test(self):
        var = gdb.lookup_symbol('ptr_to_ulong_test', None)[0]
        self.assertTrue(var is not None)
        for cpu, val in list(percpu.get_percpu_vars(var).items()):
            self.assertTrue(val.type == self.ulong_type.pointer())
            self.assertTrue(val.dereference() == cpu)

    def test_voidp_test(self):
        var = gdb.lookup_symbol('voidp_test', None)[0]
        self.assertTrue(var is not None)
        for cpu, val in list(percpu.get_percpu_vars(var).items()):
            self.assertTrue(val is not None)
            self.assertTrue(val.type == self.voidp)
            self.assertTrue(int(val) == 0xdeadbeef)

    def test_struct_test_ptr(self):
        var = gdb.lookup_symbol('ptr_to_struct_test', None)[0]
        self.assertTrue(var is not None)
        for cpu, val in list(percpu.get_percpu_vars(var).items()):
            self.assertTrue(val['x'] == cpu)
            self.assertTrue(val.type == self.test_struct.pointer())

    # This is a saved pointer to an unbound percpu var
    def test_percpu_ptr_sym(self):
        var = gdb.lookup_symbol('percpu_test', None)[0]
        self.assertTrue(var is not None)
        for cpu, val in list(percpu.get_percpu_vars(var).items()):
            self.assertTrue(val.type == self.test_struct)

    # This is a pointer to an unbound percpu var
    def test_percpu_ptr_val(self):
        var = gdb.lookup_symbol('percpu_test', None)[0].value()
        self.assertTrue(var is not None)
        for cpu, val in list(percpu.get_percpu_vars(var).items()):
            self.assertTrue(val.type == self.test_struct)

    # This is a saved pointer to an bound percpu var (e.g. normal ptr)
    def test_non_percpu_sym(self):
        var = gdb.lookup_symbol('non_percpu_test', None)[0]
        self.assertTrue(var is not None)
        with self.assertRaises(TypeError):
            x = percpu.get_percpu_var(var, 0)
        self.assertTrue(var.value()['x'] == 0)

    # This is a pointer to an bound percpu var (e.g. normal ptr)
    def test_non_percpu_ptr(self):
        var = gdb.lookup_symbol('non_percpu_test', None)[0].value()
        self.assertTrue(var is not None)
        with self.assertRaises(TypeError):
            x = percpu.get_percpu_var(var, 0)
        self.assertTrue(var['x'] == 0)
